design/multi-factor-admin-and-step-up.md

```diff
@@ -0,0 +1,613 @@
+conscientious/paranoiac to set N factors.
+By setting a single authentication factor with an authentication executions that are not
+passwords (e.g. Fido2), this system also allows passwordless login.
+
+Authentication factors are treated as a bloc in the evaluation of alternatives. That is, if in
+a flow there's "Identity provider redirector", "authentication factor 1", "authentication
+factor 2", entering the authentication factor 1 flow will automatically cause "authentication
+factor 2" to be executed after.
+
+To make things perhaps a little more clear, the current default "Browser" authentication
+would become for example:
+
+```
+Auth type
+------------------------------------------------------------
+Identity Provider Redirector
+Authentication factor (1)
+  |- Cookie
+  |- Kerberos
+  |- Username Password Form
+Authentication factor (2) [OPTIONAL]
+  |- Cookie
+  |- Kerberos
+  |- OTP Form
+```
+And if we wanted to have an mandatory 2nd factor, with either OTP or a FIDO2 configured:
+
+```
+Auth type
+------------------------------------------------------------
+Identity Provider Redirector
+Authentication factor (1)
+  |- Cookie
+  |- Kerberos
+  |- Username Password Form
+Authentication factor (2)
+  |- Cookie
+  |- Kerberos
+  |- OTP Form
+  |- FIDO-2
+```
+Potentially we could also introduce another type of authentication execution, we could call it "Authenticated on", to simplify the authenticators that bypass the authentication factors. We could then have:
+
+```
+Auth type
+------------------------------------------------------------
+Identity Provider Redirector
+Authenticated on
+  |- Cookie
+  |- Kerberos
+Authentication factor (1)
+  |- Username Password Form
+Authentication factor (2)
+  |- OTP Form
+  |- FIDO-2
+```
+
+#### Authentication categories
+
+Authentication categories allow the link between the actual execution within the flow and the
+credentials of a certain type.
+
+For example, currently there is the OTPFormAuthenticator for the browser flow and ValidateOTP for the direct grant flow.
+However, both these authentication executions will accept input from the same OTP credentials. This means that schematically, we would have:
+```
+-----------------------         ----------------------
+| OTP Credential 1    |---------|                    |    ------------------------
+-----------------------         |                    |----| OTPFormAuthenticator |
+-----------------------         | OTP Authentication |    ------------------------
+| OTP Credential 2    |---------|      Category      |    ---------------
+-----------------------         |                    |----| ValidateOTP |
+-----------------------         |                    |    ---------------
+| OTP Credential 3    |---------|                    |
+-----------------------         ----------------------
+```
+
+This schema serves as a base for all credentials and allows them to be
```

Comment:
If I understand correctly, it will be good if we can support the possibility when you have same authenticator multiple times in the authentication flow in different authentication factors.

So for example you will have the flow with the authentication factor with LoA 2, which will contain OTP authenticator with requirement of having 6 digits. And in the same flow, you will have authentication factor with LoA 3, which will contain OTP authenticator with requirement of having 8 digits.

Will this be possible with the current design?

Reply:
With the current design no, I imagined that if you have the same authentication factor within two authentication factors (or N authentication factors), then validating it once will make it valid for all the others. This was to allow a type of credential to count for N factors.

I have an idea of how to reconcile your idea with the authentication factors (or with the improved authentication flows proposed by @stianst), but do you feel that this is a necessary feature?